Restful app

artsiom-bukhautsou opened the pull request.

Commit:
DAO layer for `Developer` has been created

Commit:
gorilla/mux: get, create, update operations has been added

Commit:
-gorilla/mux handlers have been added
-logging middleware has been added

Commit:
sql script has been updated

Commit:
caching header func has been added

Commit:
-customer dao layer has been added
-customer handlers have been added
-customer table has been added to sql script

kortelyov commented on dao/entities/customer.go:
```diff
@@ -0,0 +1,9 @@
+package entities
```
use "model" or "models" better

kortelyov commented on dao/entities/customer.go:
```diff
@@ -0,0 +1,9 @@
+package entities
+
+type Customer struct {
+	Id       int64  `gorm:"column:id; not null" json:"id"`
```
"not null" is unnecessary, it's database restriction

kortelyov commented on dao/implementations/customerDAOImpl.go:
```diff
@@ -0,0 +1,62 @@
+package implementations
```
rework according layouts
https://github.com/golang-standards/project-layout

kortelyov commented on main.go:
```diff
@@ -0,0 +1,292 @@
+	writer.WriteHeader(http.StatusNoContent)
+}
+
+//--------------------------------------------------------------------------
```
wtf

Commit:
remove task.txt

Commit:
-project structure has been changed
-sql script has been updated

Commit:
Merge remote-tracking branch 'origin/restful-app' into restful-app

Commit:
code has been simplified

Commit:
header wrapper has been added

Commit:
overriding functions have been simplified

Commit:
-dao layer, handler resources for entity 'user' has been added

Commit:
my own creation of auth token has been added

Commit:
JWT middleware beta version has been added

Commit:
jwt auth has been added

Commit:
jwt auth refresh and access tokens have been added

Commit:
jwt auth with correct exp has been added

Commit:
jwt refresh auth has been updated

kortelyov commented on apis/developer.go:
```diff
@@ -0,0 +1,140 @@
+}
+
+func ServeDeveloperResource(r *mux.Router) {
+	r.Handle("/developers", jwt.VerifyMiddleware(getDevelopers)).Methods("GET")
```
VerifyMiddleware --> bad name. We should verify token/roles/permissions and not middleware

kortelyov commented on apis/user.go:
```diff
@@ -0,0 +1,99 @@
+}
+
+func ServeUserResource(r *mux.Router) {
+	r.HandleFunc("/signIn", singIn).Methods("GET")
```
no camelcase in pathes --> singin, singup

kortelyov commented on daos/customer.go:
```diff
@@ -0,0 +1,56 @@
+package daos
```
"daos" is not a golang-style name (it's came from Java), check golang-layouts repository

kortelyov commented on daos/customer.go:
```diff
@@ -0,0 +1,56 @@
+	"github.com/bukhavtsov/restful-app/models"
+)
+
+type CustomerDAO struct {
```
where's encapsulation?

kortelyov commented on models/user.go:
```diff
@@ -0,0 +1,10 @@
+package models
+
+type User struct {
```
db struct should be near at db's implementation code

Commit:
-VerifyMiddleware has been changed to VerifyPermissions
-paths have been changed
-encapsulation has been added to daos

Commit:
-package `daos` has been renamed to data
-data 'constructors' has been added
-project-layout has been updated

Commit:
-db connection has been updated to 'one-time'
-dao structures and interfaces have been updated to 'data'
-package models has been removed
